src/Engine/ExecutionState.php

```diff
@@ -96,8 +97,14 @@ public function addVariableData($name, TypedDataInterface $data) {
    * {@inheritdoc}
    */
   public function getVariable($name) {
+    // If there is no such variable, lazy-add global context variables. That
+    // way can safe time fetching global context if its not needed.
     if (!array_key_exists($name, $this->variables)) {
```

Comment:
Doing it lazy also means you need to change the hasVariable() method which only looks at the current array of variables.